controllers/cluster/state.go

```diff
@@ -0,0 +1,84 @@
+package cluster
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	etcdclient "go.etcd.io/etcd/client"
+	v1 "k8s.io/api/core/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	etcdv1alpha1 "github.com/improbable-eng/etcd-cluster-operator/api/v1alpha1"
+	"github.com/improbable-eng/etcd-cluster-operator/internal/etcd"
+)
+
+type State struct {
+	Cluster        *etcdv1alpha1.EtcdCluster
+	Members        []etcdclient.Member
+	Peers          *etcdv1alpha1.EtcdPeerList
+	DesiredService *v1.Service
+	Service        *v1.Service
+}
+
+func GetState(log logr.Logger, c client.Client, etcdapi etcd.EtcdAPI, ctx context.Context, req ctrl.Request) (*State, error) {
+	state := &State{}
+
+	var cluster etcdv1alpha1.EtcdCluster
+	err := c.Get(ctx, req.NamespacedName, &cluster)
+	if err != nil {
+		return nil, err
+	}
+
+	state.Cluster = &cluster
+	state.Cluster.Default()
```

Comment:
Isn't this dangerous? If we're trying to read the state, then defaulting things might make us miss something?